core/tracker/inclusion.go

```diff
@@ -0,0 +1,325 @@
+}
+
+// inclusionCore tracks the inclusion of submitted duties.
+// It has a simplified API to allow for easy testing.
+type inclusionCore struct {
+	mu          sync.Mutex
+	submissions []submission
+
+	missedFunc      func(context.Context, submission)
+	attIncludedFunc func(context.Context, submission, block)
+}
+
+// Submitted is called when a duty is submitted to the beacon node.
+// It adds the duty to the list of submitted duties.
+func (i *inclusionCore) Submitted(duty core.Duty, pubkey core.PubKey, data core.SignedData, delay time.Duration) error {
+	if !supported[duty.Type] {
+		return nil
+	}
+
+	var (
+		attRoot eth2p0.Root
+		err     error
+	)
+	if duty.Type == core.DutyAttester {
+		att, ok := data.(core.Attestation)
+		if !ok {
+			return errors.New("invalid attestation")
+		}
+		attRoot, err = att.HashTreeRoot()
+		if err != nil {
+			return errors.Wrap(err, "hash attestation")
+		}
+	} else if duty.Type == core.DutyAggregator {
+		agg, ok := data.(core.SignedAggregateAndProof)
+		if !ok {
+			return errors.New("invalid aggregate and proof")
+		}
+		attRoot, err = agg.Message.Aggregate.HashTreeRoot()
+		if err != nil {
+			return errors.Wrap(err, "hash aggregate")
+		}
+	} else if duty.Type == core.DutyProposer {
+		block, ok := data.(core.VersionedSignedBeaconBlock)
+		if !ok {
+			return errors.New("invalid block")
+		}
+		if eth2wrap.IsSyntheticBlock(&block.VersionedSignedBeaconBlock) {
+			return nil
+		}
+	} else if duty.Type == core.DutyBuilderProposer {
+		block, ok := data.(core.VersionedSignedBlindedBeaconBlock)
+		if !ok {
+			return errors.New("invalid blinded block")
+		}
+		if eth2wrap.IsSyntheticBlindedBlock(&block.VersionedSignedBlindedBeaconBlock) {
+			return nil
+		}
+	}
+
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.submissions = append(i.submissions, submission{
+		Duty:    duty,
+		Pubkey:  pubkey,
+		Data:    data,
+		AttRoot: attRoot,
+		Delay:   delay,
+	})
+
+	return nil
+}
+
+// Trim removes all duties that are older than the specified slot.
+// It also calls the missedFunc for any duties that have not been included.
+func (i *inclusionCore) Trim(ctx context.Context, slot int64) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	var remaining []submission
+	for _, sub := range i.submissions {
+		if sub.Duty.Slot > slot {
+			remaining = append(remaining, sub)
+			continue
+		}
+
+		// Report missed and trim
+		i.missedFunc(ctx, sub)
+	}
+	i.submissions = remaining
+}
+
+// CheckBlock checks whether the block includes any of the submitted duties.
+func (i *inclusionCore) CheckBlock(ctx context.Context, block block) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	var remaining []submission
+	for _, sub := range i.submissions {
+		switch sub.Duty.Type {
+		case core.DutyAttester, core.DutyAggregator:
+			_, ok := block.Attestations[sub.AttRoot]
+			if !ok {
+				remaining = append(remaining, sub)
+				continue
+			}
+
+			// TODO(corver): We should probably check if validator included in AggregationBits
```

Comment:
Yeah. If we wish, we can also understand if the attestation was aggregated or not by checking
```go
if NO_OF_SET_BITS(attestation.aggregation_bits) == 1 { 
  "attestation was NOT aggregated" 
}
```